src/caching_context.jl

```diff
@@ -0,0 +1,108 @@
+# we have to somehow decide before hand which things we want to recursively hash and which
```

Comment:
I ended up deciding to remove the option for users to force objects to be hashed. `WeakKeyIdDicts` do not support immutable objects (and I don't seen an obvious way to change that, given my understanding (/ lack thereof) of how GC interacts with immutable objects). Users will expect it to work for any object. Furthermore, the more I thought about this feature, the less useful it seemed. I think the better thing to do in cases where caching isn't happening is to implement something within `transformer`, so that's what I know recommend to users.

Reply:
BTW for immutable stuff, we could consider caching with a regular `IdDict` I think, like https://github.com/MakieOrg/Makie.jl/pull/3760#discussion_r1552435452

I'm not 100% sure the implications for memory buildup though, in the immutable case

Reply:
Seems like they would be essentially the same (or very similar) as build-up in the mutable case, in that any mutable structure can be held onto by some wrapping immutable structure. (E.g. lots of array types wrap `Array`).

Reply:
Maybe, if we only cache immutable data upon request of the user, that is okay? So long as we spell out the limitation that such objects now have a reference held in `CachedContext` and the user will need to release all references to that object before the cached objects can be released from memory.